[Demangler] Make Node::addChild(NULL, ...) always assert.

al45tair opened the pull request:
We need this in order to gather relevant information when this situation arises; it's clearly happening (e.g. rdar://86071019), but not often enough for it to trigger in a build with assertions enabled.

rdar://89139049

al45tair committed:
Make Node::addChild(NULL, ...) always assert.

We need this in order to gather relevant information when this situation
arises; it's clearly happening (e.g. rdar://86071019), but not often
enough for it to trigger in a build with assertions enabled.

rdar://89139049

mikeash reviewed (approved):
LGTM. Would be worth considering if there's any additional info to log here that might be helpful (like the parent's kind, child count, other contents?). Since the turnaround time on this is not going to be real quick.

al45tair committed:
Dump additional information on demangler assertion failure.

This should help to debug things when things go wrong.

rdar://89139049

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
> LGTM. Would be worth considering if there's any additional info to log here that might be helpful (like the parent's kind, child count, other contents?). Since the turnaround time on this is not going to be real quick.

Good plan. I've made the `failAssert()` function dump the node that failed the assertion, which should supply plenty of extra information.

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

mikeash commented on lib/Demangling/Demangler.cpp:
```diff
@@ -93,8 +93,12 @@ static bool isRequirement(Node::Kind kind) {
 void swift::Demangle::failAssert(const char *file, unsigned line,
                                  NodePointer node, const char *expr) {
-  fprintf(stderr, "%s:%u: assertion failed for Node %p: %s", file, line, node,
+  fprintf(stderr, "%s:%u: assertion failed for Node %p: %s\n", file, line, node,
```
We should use `swift_reportError` and `fatalError` to get these messages into the crash log too.

al45tair replied:
Hmmm. That might be slightly trickier, in that `node->dump()` isn't going to send its output to the right place. It's not a bad idea to get something in there though; I'll have a look how that all works.

al45tair replied:
It isn't quite as simple as I'd like; we can't call `swift_reportError` or `fatalError` without linking `swiftRuntime`, which not everything that includes the demangler code does.

al45tair committed:
Simplify the code slightly.

I was trying to use fatalError() here, but that is in libswiftRuntime,
which isn't linked into everything that has the demangler code in it.
The upshot is that we'd have to touch a lot of other projects to use
fatalError().

rdar://89139049

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Demangler][Runtime] Give the demangler its own error handling.

The demangling library can't use the error handling from the main runtime
because it isn't always linked with it.  However, it's useful to have
some error handling, and in particular to be able to get data into the
crash logs.

This is complicated because of the way the demangling library gets used,
the upshot of which is that I've had to add a second object library just
for libswiftCore's use, so that the demangler will use the runtime's
error handling functions when present, and fall back on its own when
they aren't.

rdar://89139049

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Build][Demangler] Enable crash reporter integration on Darwin.

Crash reporter integration was only enabled for iOS.  Enable it for
any Darwin platform, but disable it for the minimal build.

Also fix up a couple of issues that popped up when it was enabled.

rdar://89139049

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please test
